### PR TITLE
Timestamp issues

### PR DESCRIPTION
**Pull request by R-maan**
Fix malformed Ion timestamps.
Fixes include:
- Year, month, day must be larger than 0 (minimum date is 0001-01-01)
- Months cannot be larger than 12.
- Days, depending on which month, should be either 28,29,30 or 31.
- All different valid formats should be parsed correctly.
- Parse leap years properly.

This fix **does not** include:
- Timezones 
- Fractional seconds
- Binary timestamp

### CODE REVIEW

File: ion/textutils.go [hunk 1]
```diff
@@ -309,12 +309,12 @@ func parseTimestamp(val string) (time.Time, error) {
 	}
 
 	year, err := strconv.ParseInt(val[:4], 10, 32)
-	if err != nil {
+	if err != nil || year < 1 {
```
Comment (matthewh-BQ) [inline ion/textutils.go:330]: Are we able to do all the verification upfront at the start of the function? I would suggest a helper function that does these checks upfront and keep the standard Golang error handling "clean".

On the flipside, if we look at https://golang.org/pkg/time/, is it possible to create a layout and use the time parser to do all the heavy lifting for us? The len checks seems a bit manual and if I understand the layout function correctly, I think we could define it as "yyyy-mm-ddThh:mm:ss" and hope the parse will catch any of those edge cases for us.

Comment (R-maan) [reply ion/textutils.go:330]: Not sure if that is the best option/most efficient. Having a separate functions requires to process the text twice.

This function is responsible to convert the raw string into Timestamp, reading characters one by one and decides what to do next. For example the 5th character can be "T", "-" or "other characters than these 2". Then we will have 3 different paths, 2 of them valid one invalid. 
Then we can only decide based on what we have already read and what comes next. For instance "20002T" is valid where "2002-" is valid ONLY if it follows by 2 digits.

Imagine the following scenarios:
- The string is "2002-03". We can make sure the first 4 characters are a digit larger than 1. Same for chars [6:8]. But the entire text is not valid.
- A bigger example is something like "2002-03-03T12:34:45.5555+12:12". We do not know the format beforehand (as this is a raw string) and at any point we can have either a) an error or b) a valid timestamp. Validating it before trying to parse it requires processing the entire text first. And then another time to construct the time.

About the use of layout, the code already is using it, but we must know the exact format our text can translate to. Some examples of using layouts are [here ](https://github.com/Armanbqt/ion-go/pull/2/files#diff-81c3138632e57790bcb2cd643f1bbdb3R400)and [here](https://github.com/Armanbqt/ion-go/pull/2/files#diff-81c3138632e57790bcb2cd643f1bbdb3R367). In order to know which layout we can/should use based on our input, we should know the text. We do not want to pass "2002-02T" to ```time.parse()``` with "yyyy-mm-ddThh:mm:ss"

Comment (matthewh-BQ) [reply ion/textutils.go:330]: Talked offline, thanks for the explanation. Please ignore. Thanks.

**Comment by R-maan** [comment]
The only remaining timestamp tests which are not passing for loadBad() are:
- timestampHourWithoutMinute.10n
- timestampLenTooLarge.10n

**Pull request closed without merging**